Rick and Morty character screen

Adam0091 opened the pull request:
1. Task: 
  Реализовать приложение Rick and Morty на ReactJS
  Стек - ReactJS, GraphQL, (Styled Component/CSS Modules + scss), react-router-dom
  Дизайн - https://www.figma.com/file/J3BzEpZLasRlhP0836rnk9/Rick-and-Morty-(web-responsive)-(Community)?node-id=0%3A3
  (Дизайн для примера, строго к нему не привязываемся)
  API - https://rickandmortyapi.com/documentation 

  Документация GraphQL - https://graphql.org/ 
  Гайд по apollo - https://www.apollographql.com/tutorials
  Генерация типов - https://github.com/dotansimha/graphql-code-generator
2. Screenshot: 
![image](https://user-images.githubusercontent.com/55671230/182227474-096f4937-55d6-4841-b1c7-f4249be13ea5.png)

3. Deploy: https://adam0091.github.io/rick-and-morty/
4. Done 1.08.2022

Adam0091 committed:
feat: add eslint and alias

Adam0091 committed:
feat: add header and characters components

Adam0091 committed:
feat: add Header component

Adam0091 committed:
feat: add Footer component

Adam0091 committed:
feat: add constants file

Adam0091 committed:
feat: add Container style component

Adam0091 committed:
feat: add CharacterCard component

Adam0091 committed:
feat: add CharactersList component

Adam0091 committed:
feat: add Apollo provider

Adam0091 committed:
feat: add new alies

Adam0091 committed:
fix: del comment and change title

Adam0091 committed:
feat: add complete CharactersPage

Adam0091 committed:
fix: change rules for eslint

Adam0091 committed:
feat: add mui, apollo, props-types, graphql

Adam0091 committed:
refactor: use lint

Adam0091 committed:
fix: fix errors from console

Adam0091 committed:
feat: add CharactersFilters component

Adam0091 committed:
feat: add FillterButton component

Adam0091 committed:
feat: add filterIcon for FiltersButton

Adam0091 committed:
feat: add FiltersInputSearch

Adam0091 committed:
feat: add FiltersSelects component

Adam0091 committed:
fix: delete Style Component and add ClassName

Adam0091 committed:
fix: bug with adaptive

Adam0091 committed:
refactor: use lint

Adam0091 committed:
fix: add className

Adam0091 committed:
feat: add Burger component

Adam0091 committed:
feat: add Select component

Adam0091 committed:
feat: add Shadow component

Adam0091 committed:
fix: delete icon

Adam0091 committed:
fix: breakdown into components

Adam0091 committed:
feat: add adaptive fro Header

Adam0091 committed:
fix: delete canScroll

Adam0091 committed:
fix: use lint

Impeqq commented on src/components/CharacterCard/CharacterCard.jsx:
```diff
@@ -0,0 +1,26 @@
+
+  return (
+    <CharacterWrapper>
+      <CharacterImage image={image} />
```
почему ты не используешь тег img? можно было бы написать ему alt={name} это лучше чем background-image

Adam0091 replied:
Это косяк. Тег img действительно здесь уместнее, так как это является важной частью контента и конечно alt здесь тоже нужен для SEO

Impeqq commented on src/components/CharacterCard/CharacterCard.style.js:
```diff
@@ -0,0 +1,65 @@
+  padding: 12px 16px;
+
+  .character__name {
+    font-family: "Roboto";
```
почему бы не написать font-family в глобальных стилях? вижу 8 раз в коде пишешь это свойство, можно написать в одном месте один раз

Adam0091 replied:
Честно, редко обращал на это внимание, но если присмотреться то это противоречит принципу DRI. Впредь буду внимательнее к дублированию стилей

Impeqq commented on src/components/CharacterCard/index.js:
```diff
@@ -0,0 +1 @@
+export { default } from "./CharacterCard";
```
давай не использовать export default, везде пишем export * from..

Adam0091 replied:
Хорошо. 
Только изучил этот вопрос, что лучше и почему. И сделал для себя открытие)
Как я понял это увеличивает явность кода, то что ты видишь тем и является и не будет такого что, к примеру класс Apple станет Orange.
И при использование поиска, будет точно известно что это именно тот модуль, который нужен, а не переименованная его версия.
Плюсом сразу предупреждает если сущности не существует. 
В статьях часто писали ещё про "прогулки по коду", но не совсем понимаю, как именованные экспорты с этим помогают.

Impeqq commented on src/components/CharactersFilters/FiltersButton/FiltersButton.jsx:
```diff
@@ -0,0 +1,28 @@
+
+const FiltersButton = ({ isActive, setIsActive }) => {
+  return (
+    <WrapperFiltersButton onClick={() => setIsActive(!isActive)}>
```
когда у тебя компонент возвращает только шаблон можно просто писать const FiltersButton = ({ isActive, setIsActive }) => {(<WrapperFiltersButton.......

давай сделай лучше хендлер и там setIsActive((isActive) => !isActive)

Adam0091 replied:
Да, конечно, можно сразу возвращать компонент без return. Я всегда думал, что это лишает возможности быстро вносить функции при дорабатывание логики компонента или например для выноса обработки событий в handle, или ошибаюсь?

Impeqq commented on src/components/CharactersFilters/FiltersInputSearch/FiltersInputSearch.jsx:
```diff
@@ -0,0 +1,30 @@
+
+import { WrapperInputSearch } from "./FiltersInputSearch.style";
+
+import searchIcon from "./img/searchIcon.svg";
```
создай папку src/assets/images там будут все картинки jpg/svg/....    удобнее хранить всё в одном месте а не по разным папкам проекта

Adam0091 replied:
В своих предыдущих проектах тоже хранил всё в assets/images, потом как то увидел подомную архитектуру с хранением img в компонентах и решил попробовать. А раз уж опытный разработчик мне говорит что так лучше, грех не довериться и оставить сей эксперимент)

Impeqq commented on src/components/CharactersFilters/FiltersInputSearch/FiltersInputSearch.jsx:
```diff
@@ -0,0 +1,30 @@
+        type="text"
+        placeholder="Filter by name..."
+        onChange={(e) => {
+          setFilterOptions({
```
всегда используй setFilterOptions((filterOptions) => ({...filterOptions.....  не придется передавать лишний проп и стейт будет всегда актуальный, а так есть возможность что-то сломать

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.jsx:
```diff
@@ -0,0 +1,117 @@
+
+  const handleSelect = (value, type) => {
+    if (window.innerWidth > 767) {
+      setValues({
```
и тут

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.jsx:
```diff
@@ -0,0 +1,117 @@
+        ...values,
+        [type]: value,
+      });
+      setFilterOptions({
```
и тут

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.jsx:
```diff
@@ -0,0 +1,117 @@
+        [type]: value,
+      });
+    } else {
+      setValues({
```
и тут

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.jsx:
```diff
@@ -0,0 +1,117 @@
+  };
+
+  const handleApplyButton = () => {
+    setIsActive(!isActive);
```
и тут

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.jsx:
```diff
@@ -0,0 +1,117 @@
+
+  const handleApplyButton = () => {
+    setIsActive(!isActive);
+    setFilterOptions({
```
и тут

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.jsx:
```diff
@@ -0,0 +1,117 @@
+  };
+
+  const onCloseSelects = () => {
+    setIsActive(!isActive);
```
и тут

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.jsx:
```diff
@@ -0,0 +1,117 @@
+
+  const onCloseSelects = () => {
+    setIsActive(!isActive);
+    setValues({
```
и тут, плюс {... filterOptions} одно и то же с тем что ты написал

Adam0091 replied:
Не совсем, в filterOptions есть поле name, за которок отвечает FilterInputSearch
![image](https://user-images.githubusercontent.com/55671230/182365805-d751cdc4-bdf2-4bf9-aab1-f78781f800cc.png)
![image](https://user-images.githubusercontent.com/55671230/182365898-5ae72d41-83c3-4cc3-b42a-7c876523f150.png)
Поэтому я решил расписать, но походу только вызвал путаницу)

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.style.js:
```diff
@@ -0,0 +1,96 @@
+    display: ${({ isActive }) => (isActive ? "flex" : "none")};
+    position: fixed;
+    width: 100%;
+    box-sizing: border-box;
```
box-sizing же и так стоит в глобальных стилях?

Impeqq commented on src/components/CharactersFilters/FiltersSelects/FiltersSelects.style.js:
```diff
@@ -0,0 +1,96 @@
+  height: 14px;
+  width: 14px;
+
+  padding-right: 5px;
```
этот padding ничего не делает 
<img width="693" alt="image" src="https://user-images.githubusercontent.com/67609910/182243281-cf0f5401-a958-4c36-acc6-440c212c2f01.png">

Adam0091 replied:
Я озадачен 🤔. box-sizing: border-box сработал там где не должен был, учитывая что в глобальных стилях он не сработал, из-за неправильного объявления, получается должно было сработать значение по умолчанию content-box

Impeqq commented on src/components/Header/Burger/Burger.jsx:
```diff
@@ -0,0 +1,18 @@
+const Burger = ({ isActive, setIsActive }) => {
+  return (
+    <BurgerContainer isActive={isActive} onClick={() => setIsActive(!isActive)}>
+      <span></span>
```
<span />, и выше стейт колбеком ставим, как в других местах

Impeqq commented on src/index.js:
```diff
@@ -1,10 +1,22 @@
+import { normalize } from "styled-normalize";
+
+const Global = createGlobalStyle`
+  box-sizing: border-box;
```
сейчас это не работает, сделай * { box-sizing: border-box; }

Adam0091 replied:
Это объясняет почему местами приходилось использовать костыль)

Adam0091 committed:
fix: migration to typescript

Adam0091 committed:
refactor: use lint

Adam0091 committed:
fix: removal of duplication of font-family,
transition from default export to named export,
in CharacterCard replace background image with teg img,
delete PropsTypes

Adam0091 committed:
fix: moving images to assets/images

Adam0091 committed:
fix: fixed global styles

Adam0091 committed:
fix: delete unnecessary props,
fixed state change in CharactersInputSearch

Adam0091 committed:
fix: delete unnecessary styles(box-sizing)
and add box-sizing: content-box in Container

Adam0091 committed:
fix: fixed setState and add handles

Adam0091 committed:
fix: delete console.log

Adam0091 committed:
feat: add types

Adam0091 committed:
feat: add block scroll
